File: api_sample/typescript/rest-list-cameras-cloud-user/test_rest_cloud_sample.ts
// Copyright 2018-present Network Optix, Inc. Licensed under MPL 2.0: www.mozilla.org/MPL/2.0/
/**
 * Offline tests for rest_cloud_sample.ts. No network, no account needed.
 *
 * Run from this folder:  node --test test_rest_cloud_sample.ts
 */

import assert from "node:assert/strict";
import test from "node:test";

import type { Camera } from "../nx-types.ts";

import {
  NxCloudSiteClient,
  formatCamerasTable,
  resolveConfig,
  AuthError,
  ApiError,
  type ClientOptions,
} from "./rest_cloud_sample.ts";

interface FakeResponseSpec {
  status?: number;
  json?: unknown;
  text?: string;
  headers?: Record<string, string>;
}

function makeResponse({
  status = 200,
  json = null,
  text = "",
  headers = {},
}: FakeResponseSpec = {}): Response {
  const lower: Record<string, string> = {};
  for (const [k, v] of Object.entries(headers)) {
    lower[k.toLowerCase()] = v;
  }
  return {
    status,
    ok: status < 400,
    headers: { get: (name: string) => lower[name.toLowerCase()] ?? null },
    async json() {
      if (json === null) throw new Error("no json");
      return json;
    },
    async text() {
      return text;
    },
  } as unknown as Response;
}

interface RecordedGet {
  url: string;
  headers: Record<string, string> | undefined;
  redirect: string | undefined;
}

interface FakeFetchSpec {
  post?: Response | null;
  gets?: Response[];
  del?: Response | null;
}

interface FetchCalls {
  post: { url: string; body: Record<string, unknown> | null } | null;
  gets: RecordedGet[];
  deleteUrl: string | null;
  deleteCalls: number;
}

type FakeFetch = typeof globalThis.fetch & { calls: FetchCalls };

function fakeFetch({ post = null, gets = [], del = null }: FakeFetchSpec = {}): FakeFetch {
  const getQ = [...gets];
  const calls: FetchCalls = { post: null, gets: [], deleteUrl: null, deleteCalls: 0 };
  const impl = (async (url: string | URL | Request, options: RequestInit = {}) => {
    const method = (options.method || "GET").toUpperCase();
    const urlStr = String(url);
    if (method === "POST") {
      calls.post = {
        url: urlStr,
        body: options.body ? (JSON.parse(options.body as string) as Record<string, unknown>) : null,
      };
      return post as Response;
    }
    if (method === "DELETE") {
      calls.deleteUrl = urlStr;
      calls.deleteCalls += 1;
      return del as Response;
    }
    calls.gets.push({
      url: urlStr,
      headers: options.headers as Record<string, string> | undefined,
      redirect: options.redirect,
    });
    return getQ.shift() as Response;
  }) as FakeFetch;
  impl.calls = calls;
  return impl;
}

const SYS = "11111111-2222-3333-4444-555555555555";

function makeClient(
  fetchOpts: FakeFetchSpec = {},
  opts: ClientOptions = {},
): { client: NxCloudSiteClient; f: FakeFetch } {
  const f = fakeFetch(fetchOpts);
  const client = new NxCloudSiteClient("https://nxvms.com", "me@x.com", "pw", SYS, {
    fetchImpl: f,
    ...opts,
  });
  return { client, f };
}

// ---------------------------------------------------------------------------
// login(): the token MUST carry the cloudSystemId scope
// ---------------------------------------------------------------------------

test("login includes the site scope", async () => {
  const { client, f } = makeClient({ post: makeResponse({ json: { access_token: "nxcdb-t" } }) });
  const token = await client.login();
  assert.equal(token, "nxcdb-t");
  assert.equal(f.calls.post!.url, "https://nxvms.com/cdb/oauth2/token");
  assert.equal(f.calls.post!.body!.scope, `cloudSystemId=${SYS}`);
  assert.equal(f.calls.post!.body!.client_id, "3rdParty");
});

test("login adds the mfa code", async () => {
  const { client, f } = makeClient(
    { post: makeResponse({ json: { access_token: "t" } }) },
    { mfaCode: "999111" },
  );
  await client.login();
  assert.equal(f.calls.post!.body!.mfaCode, "999111");
});

test("login rejected raises AuthError", async () => {
  const { client } = makeClient({ post: makeResponse({ status: 403, text: "no" }) });
  await assert.rejects(() => client.login(), AuthError);
});

// ---------------------------------------------------------------------------
// relay url + listCameras() — uses the latest /rest/v4 via the relay
// ---------------------------------------------------------------------------

test("relayUrl is built from the site id", () => {
  const { client } = makeClient();
  assert.equal(client.relayUrl, `https://${SYS}.relay.vmsproxy.com`);
});

test("listCameras uses the relay v4 path, bearer, and manual redirect", async () => {
  const payload: Camera[] = [{ id: "c1", name: "Lobby", status: "Online", model: "Axis" }];
  const { client, f } = makeClient({ gets: [makeResponse({ json: payload })] });
  client.token = "nxcdb-t";
  const cams = await client.listCameras();
  assert.equal(cams[0]!.name, "Lobby");
  const call = f.calls.gets[0] as RecordedGet;
  assert.equal(call.url, `https://${SYS}.relay.vmsproxy.com/rest/v4/devices`);
  assert.equal(call.headers?.Authorization, "Bearer nxcdb-t");
  assert.equal(call.redirect, "manual"); // we follow redirects ourselves
});

test("listCameras follows a 307 with the bearer preserved", async () => {
  const node = "https://node7.relay.vmsproxy.com/rest/v4/devices";
  const payload: Camera[] = [{ id: "c1", name: "Lobby", status: "Online", model: "Axis" }];
  const { client, f } = makeClient({
    gets: [
      makeResponse({ status: 307, headers: { Location: node } }),
      makeResponse({ json: payload }),
    ],
  });
  client.token = "nxcdb-t";

  const cams = await client.listCameras();

  assert.equal(f.calls.gets.length, 2);
  // Crucially, the bearer header is re-attached on the redirected request.
  assert.equal((f.calls.gets[1] as RecordedGet).url, node);
  assert.equal((f.calls.gets[1] as RecordedGet).headers?.Authorization, "Bearer nxcdb-t");
  assert.equal(cams[0]!.name, "Lobby");
});

test("listCameras redirect without a Location header raises", async () => {
  const { client } = makeClient({ gets: [makeResponse({ status: 307, headers: {} })] });
  client.token = "t";
  await assert.rejects(() => client.listCameras(), ApiError);
});

test("listCameras gives up after too many redirects", async () => {
  const node = "https://node7.relay.vmsproxy.com/rest/v4/devices";
  const { client } = makeClient({
    gets: Array.from({ length: 7 }, () => makeResponse({ status: 307, headers: { Location: node } })),
  });
  client.token = "t";
  await assert.rejects(() => client.listCameras(), ApiError);
});

test("listCameras unwraps a reply envelope", async () => {
  const { client } = makeClient({ gets: [makeResponse({ json: { reply: [{ id: "c1", name: "L" }] } })] });
  client.token = "t";
  const cams = await client.listCameras();
  assert.equal(cams[0]!.name, "L");
});

test("listCameras without login raises", async () => {
  const { client } = makeClient();
  await assert.rejects(() => client.listCameras(), ApiError);
});

// ---------------------------------------------------------------------------
// logout() deletes the token ON THE CLOUD
// ---------------------------------------------------------------------------

test("logout deletes the token on the cloud", async () => {
  const { client, f } = makeClient({ del: makeResponse({ status: 204 }) });
  client.token = "nxcdb-t";
  await client.logout();
  assert.equal(f.calls.deleteUrl, "https://nxvms.com/cdb/oauth2/token/nxcdb-t");
  assert.equal(client.token, null);
});

// ---------------------------------------------------------------------------
// config + table
// ---------------------------------------------------------------------------

test("config reads the site id (env beats file)", () => {
  const args = { cloudHost: null, user: null, password: null, siteId: null, mfaCode: null };
  const config = resolveConfig(
    args,
    { NX_CLOUD_SITE_ID: "file-sys" },
    { NX_CLOUD_SITE_ID: "env-sys" } as NodeJS.ProcessEnv,
  );
  assert.equal(config.siteId, "env-sys");
});

test("formatCamerasTable renders rows and the empty case", () => {
  assert.ok(formatCamerasTable([{ name: "Lobby" }]).includes("Lobby"));
  assert.ok(formatCamerasTable([]).includes("No cameras"));
});

File: api_sample/python/rest-list-cameras-cloud-user/test_rest_cloud_sample.py
# Copyright 2018-present Network Optix, Inc. Licensed under MPL 2.0: www.mozilla.org/MPL/2.0/
"""
Offline tests for rest_cloud_sample.py. No network, no account needed.

Run from this folder:  pytest -v
"""

import argparse

import pytest

import rest_cloud_sample as sample


class FakeResponse:
    def __init__(self, status_code=200, json_data=None, text="", headers=None):
        self.status_code = status_code
        self._json = json_data
        self.text = text
        self.headers = headers or {}

    @property
    def ok(self):
        return self.status_code < 400

    def json(self):
        if self._json is None:
            raise ValueError("no json")
        return self._json


class FakeSession:
    """Serves queued GET responses in order (for redirect-chasing tests);
    POST/DELETE each return one fixed response."""

    def __init__(self, post=None, get=None, gets=None, delete=None):
        self.verify = None
        self._post, self._delete = post, delete
        self._gets = list(gets) if gets is not None else ([get] if get is not None else [])
        self.post_url = self.post_json = None
        self.get_calls = []  # (url, headers, allow_redirects)
        self.delete_url = None
        self.delete_calls = 0

    def post(self, url, json=None, timeout=None):
        self.post_url, self.post_json = url, json
        return self._post

    def get(self, url, headers=None, timeout=None, allow_redirects=None):
        self.get_calls.append((url, headers, allow_redirects))
        return self._gets.pop(0)

    def delete(self, url, headers=None, timeout=None):
        self.delete_url = url
        self.delete_calls += 1
        return self._delete

    # -- back-compat helpers used by the pre-existing single-GET tests --
    @property
    def get_url(self):
        return self.get_calls[-1][0] if self.get_calls else None

    @property
    def get_headers(self):
        return self.get_calls[-1][1] if self.get_calls else None

    @property
    def get_kwargs(self):
        return {"allow_redirects": self.get_calls[-1][2]} if self.get_calls else None


SYS = "11111111-2222-3333-4444-555555555555"


def make_client(**kw):
    defaults = dict(cloud_host="https://nxvms.com", user="me@x.com",
                    password="pw", site_id=SYS)
    defaults.update(kw)
    return sample.NxCloudSiteClient(**defaults)


# ---------------------------------------------------------------------------
# login(): the token MUST carry the cloudSystemId scope
# ---------------------------------------------------------------------------

def test_login_includes_system_scope():
    session = FakeSession(post=FakeResponse(200, {"access_token": "nxcdb-t"}))
    client = make_client(session=session)

    token = client.login()

    assert token == "nxcdb-t"
    assert session.post_url == "https://nxvms.com/cdb/oauth2/token"
    assert session.post_json["scope"] == f"cloudSystemId={SYS}"
    assert session.post_json["client_id"] == "3rdParty"


def test_login_adds_mfa_code():
    session = FakeSession(post=FakeResponse(200, {"access_token": "t"}))
    make_client(session=session, mfa_code="999111").login()
    assert session.post_json["mfaCode"] == "999111"


def test_login_rejected_raises_autherror():
    session = FakeSession(post=FakeResponse(403, text="no"))
    with pytest.raises(sample.AuthError):
        make_client(session=session).login()


# ---------------------------------------------------------------------------
# relay url + list_cameras()
# ---------------------------------------------------------------------------

def test_relay_url():
    assert make_client().relay_url == f"https://{SYS}.relay.vmsproxy.com"


def test_list_cameras_uses_relay_and_bearer():
    payload = [{"id": "c1", "name": "Lobby", "status": "Online", "model": "Axis"}]
    session = FakeSession(get=FakeResponse(200, payload))
    client = make_client(session=session)
    client.token = "nxcdb-t"

    cams = client.list_cameras()

    assert cams[0]["name"] == "Lobby"
    assert session.get_url == f"https://{SYS}.relay.vmsproxy.com/rest/v4/devices"
    assert session.get_headers["Authorization"] == "Bearer nxcdb-t"
    assert session.get_kwargs["allow_redirects"] is False


def test_list_cameras_unwraps_reply_envelope():
    session = FakeSession(get=FakeResponse(200, {"reply": [{"id": "c1", "name": "L"}]}))
    client = make_client(session=session)
    client.token = "t"
    assert client.list_cameras()[0]["name"] == "L"


def test_list_cameras_without_login_raises():
    with pytest.raises(sample.ApiError):
        make_client(session=FakeSession()).list_cameras()


def test_list_cameras_follows_307_with_bearer_preserved():
    # First GET -> 307 to the serving node; second GET -> the data.
    node = "https://node7.relay.vmsproxy.com/rest/v4/devices"
    session = FakeSession(gets=[
        FakeResponse(307, headers={"Location": node}),
        FakeResponse(200, [{"id": "c1", "name": "Lobby"}]),
    ])
    client = make_client(session=session)
    client.token = "nxcdb-t"

    cams = client.list_cameras()

    assert len(session.get_calls) == 2
    # Crucially, the bearer header is re-attached on the redirected request.
    assert session.get_calls[1][0] == node
    assert session.get_calls[1][1]["Authorization"] == "Bearer nxcdb-t"
    assert session.get_calls[1][2] is False  # still not auto-following
    assert cams[0]["name"] == "Lobby"


def test_list_cameras_redirect_without_location_raises():
    session = FakeSession(gets=[FakeResponse(307, headers={})])
    client = make_client(session=session)
    client.token = "t"
    with pytest.raises(sample.ApiError):
        client.list_cameras()


def test_list_cameras_too_many_redirects_raises():
    node = "https://node7.relay.vmsproxy.com/rest/v4/devices"
    # One more 307 than MAX_REDIRECTS allows -> should give up.
    session = FakeSession(
        gets=[FakeResponse(307, headers={"Location": node})] * (sample.MAX_REDIRECTS + 1)
    )
    client = make_client(session=session)
    client.token = "t"
    with pytest.raises(sample.ApiError):
        client.list_cameras()


# ---------------------------------------------------------------------------
# logout() deletes the token ON THE CLOUD
# ---------------------------------------------------------------------------

def test_logout_deletes_token_on_cloud():
    session = FakeSession(delete=FakeResponse(204))
    client = make_client(session=session)
    client.token = "nxcdb-t"

    client.logout()

    assert session.delete_url == "https://nxvms.com/cdb/oauth2/token/nxcdb-t"
    assert client.token is None


# ---------------------------------------------------------------------------
# config
# ---------------------------------------------------------------------------

def test_config_reads_site_id(monkeypatch):
    monkeypatch.setenv("NX_CLOUD_SITE_ID", "env-sys")
    args = argparse.Namespace(cloud_host=None, user=None, password=None,
                              site_id=None, mfa_code=None)
    config = sample.resolve_config(args, {"NX_CLOUD_SITE_ID": "file-sys"})
    assert config["site_id"] == "env-sys"


def test_format_cameras_table():
    out = sample.format_cameras_table([{"name": "Lobby"}])
    assert "Lobby" in out
    assert "No cameras" in sample.format_cameras_table([])
